File: reef-tang/tang/src/main/java/com/microsoft/tang/util/AbstractMonotonicMultiMap.java
/**
 * Copyright (C) 2014 Microsoft Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.microsoft.tang.util;

import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

public abstract class AbstractMonotonicMultiMap<K,V> implements Collection<Entry<K,V>> {
  protected Map<K, Set<V>> map;
  private int size = 0;
  public AbstractMonotonicMultiMap(Map<K, Set<V>> map) {
    this.map = map;
  }
  public void put(K key, V v) {
    Set<V> vals = map.get(key);
    if(vals == null) {
      vals = new MonotonicHashSet<V>();
      map.put(key,vals);
    }
    vals.add(v);
    size++;
  }
  public Set<K> keySet() {
    return map.keySet();
  }
  public Set<V> getValuesForKey(K key) {
    Set<V> ret = map.get(key);
    if(ret == null) {
      return new MonotonicHashSet<V>();
    } else {
      return ret;
    }
  }
  public boolean contains(K key, V v) {
    Set<V> vals = map.get(key);
    if(vals != null) {
      return vals.contains(v);
    }
    return false;
  }
  @Override
  public boolean add(Entry<K, V> e) {
    put(e.getKey(), e.getValue());
    return true;
  }
  @Override
  public boolean addAll(Collection<? extends Entry<K, V>> c) {
    boolean ret = false;
    for(Entry<K,V> e : c) {
      add(e);
      ret = true;
    }
    return ret;
  }
  @Override
  public void clear() {
    throw new UnsupportedOperationException("MonotonicMultiMap cannot be cleared!");
  }
  @SuppressWarnings("unchecked")
  @Override
  public boolean contains(Object o) {
    Entry<?,?> e = (Entry<?,?>)o;
    return contains((K)e.getKey(), (V)e.getValue());
  }
  @Override
  public boolean containsAll(Collection<?> c) {
    for(Object o : c) {
      if(!contains(o)) { return false; }
    }
    return true;
  }
  @Override
  public boolean isEmpty() {
    return size == 0;
  }
  @Override
  public Iterator<Entry<K, V>> iterator() {
    final Iterator<Entry<K, Set<V>>> it = map.entrySet().iterator();
    return new Iterator<Entry<K,V>>() {
      Iterator<V> cur;
      K curKey;
      @Override
      public boolean hasNext() {
        return it.hasNext() || (cur != null && cur.hasNext());
      }

      @Override
      public Entry<K,V> next() {
        if(cur == null) {
          if(it.hasNext()) {
            Entry<K,Set<V>> e = it.next();
            curKey = e.getKey();
            cur = e.getValue().iterator();
          }
        }
        final K k = curKey;
        final V v = cur.next();
        if(!cur.hasNext()) { cur = null; }
        
        return new Entry<K,V>() {

          @Override
          public K getKey() {
            return k;
          }

          @Override
          public V getValue() {
            return v;
          }

          @Override
          public V setValue(V value) {
            throw new UnsupportedOperationException();
          }
          
        };
      }

      @Override
      public void remove() {
        throw new UnsupportedOperationException();
      }
      
    };
  }
  public Set<V> values() {
    Set<V> s = new HashSet<>();
    for(Entry<K,V> e : this) {
      s.add(e.getValue());
    }
    return s;
  }
  @Override
  public boolean remove(Object o) {
    throw new UnsupportedOperationException("MonotonicMultiMap does not support non-monotonic method remove!");
  }
  @Override
  public boolean removeAll(Collection<?> c) {
    throw new UnsupportedOperationException("MonotonicMultiMap does not support non-monotonic method removeAll!");
  }
  @Override
  public boolean retainAll(Collection<?> c) {
    throw new UnsupportedOperationException("MonotonicMultiMap does not support non-monotonic method retainAll!");
  }
  @Override
  public int size() {
    return size;
  }
  @Override
  public Entry<K,V>[] toArray() {
    throw new UnsupportedOperationException("No toArray() for MonotonicMulitMap (yet)");
  }
  @Override
  public <T> T[] toArray(T[] a) {
    throw new UnsupportedOperationException("No toArray() for MonotonicMulitMap (yet)");
  }
  public boolean containsKey(K k) {
    if(map.containsKey(k)) {
      return !getValuesForKey(k).isEmpty();
    } else {
      return false;
    }
  }
}

File: lang/cs/Org.Apache.REEF.Tang/Util/AbstractMonotonicMultiMap.cs
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Org.Apache.REEF.Utilities.Logging;

namespace Org.Apache.REEF.Tang.Util
{
    public abstract class AbstractMonotonicMultiMap<K, V> : ICollection<KeyValuePair<K, V>>
    {
        private static readonly Logger LOGGER = Logger.GetLogger(typeof(AbstractMonotonicMultiMap<K, V>));

        private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);

        private readonly IDictionary<K, ISet<V>> map;

        private int size = 0;

        public AbstractMonotonicMultiMap(IDictionary<K, ISet<V>> map)
        {
            this.map = map;
        }

        public ICollection<K> Keys
        {
            get { return map.Keys; }
        }

        public int Count
        {
            get { return size; }
        }

        public bool IsReadOnly
        {
            get { throw new NotImplementedException(); }
        }

        public void Add(K key, V val)
        {
            _lock.EnterWriteLock();
            try
            {
                ISet<V> vals;
                map.TryGetValue(key, out vals);

                if (vals == null)
                {
                    vals = new MonotonicHashSet<V>();
                    map.Add(key, vals);
                }
                vals.Add(val);
                size++;
            }
            finally
            {
                if (_lock.IsWriteLockHeld)
                {
                    _lock.ExitWriteLock();
                }
            }
        }

        public ISet<V> GetValuesForKey(K key)
        {
            _lock.EnterReadLock();
            try
            {
                ISet<V> ret;
                map.TryGetValue(key, out ret);
                if (ret == null)
                {
                    return new MonotonicHashSet<V>();
                }
                return ret;
            }
            finally
            {
                if (_lock.IsReadLockHeld)
                {
                    _lock.ExitReadLock();
                }
            }
        }

        public bool Contains(K key, V val)
        {
            _lock.EnterReadLock();
            try
            {
                ISet<V> vals;
                map.TryGetValue(key, out vals);

                if (vals != null)
                {
                    return vals.Contains(val);
                }
                return false;
            }
            finally
            {
                if (_lock.IsReadLockHeld)
                {
                    _lock.ExitReadLock();
                }
            }
        }

        public bool Add(KeyValuePair<K, V> e)
        {
            Add(e.Key, e.Value);
            return true;
        }

        public bool AddAll(ICollection<KeyValuePair<K, V>> c) // where T : KeyValuePair<K, V>
        {
            bool ret = false;
            foreach (KeyValuePair<K, V> e in c) 
            {
                Add(e);
                ret = true;
            }
            return ret;
        }

        public void Clear()
        {
            throw new NotSupportedException("MonotonicMultiMap cannot be cleared!");
        }

        public bool Contains(object o) 
        {
            KeyValuePair<K, V> e = (KeyValuePair<K, V>)o;
            return Contains((K)e.Key, (V)e.Value);
        }

        public bool ContainsAll<T>(ICollection<T> c) 
        {
            foreach (object o in c) 
            {
                if (!Contains(o)) 
                { 
                    return false; 
                }
            }
            return true;
        }

        public bool IsEmpty()
        {
            return size == 0;
        }

        public ISet<V> Values() 
        {
            _lock.EnterReadLock();

            try
            {
                ISet<V> s = new HashSet<V>();
                foreach (KeyValuePair<K, V> e in this)
                {
                    s.Add(e.Value);
                }
                return s;
            }
            finally
            {
                if (_lock.IsReadLockHeld)
                {
                    _lock.ExitReadLock();
                }
            }
        }

        public int Size()
        {
            return size;
        }

        public bool ContainsKey(K k)
        {
            _lock.EnterReadLock();

            try
            {
                if (map.ContainsKey(k))
                {
                    return GetValuesForKey(k).Count != 0;
                }
                return false;
            }
            finally 
            {                
                if (_lock.IsReadLockHeld)
                {
                    _lock.ExitReadLock();
                }
            }
        }

        void ICollection<KeyValuePair<K, V>>.Add(KeyValuePair<K, V> item)
        {
            throw new NotImplementedException();
        }

        public void CopyTo(KeyValuePair<K, V>[] array, int arrayIndex)
        {
            throw new NotImplementedException();
        }

        public bool Remove(KeyValuePair<K, V> item)
        {
            throw new NotImplementedException();
        }

        public bool Contains(KeyValuePair<K, V> item)
        {
            throw new NotImplementedException();
        }

        public IEnumerator<KeyValuePair<K, V>> GetEnumerator()
        {
            _lock.EnterReadLock();

            try
            {
                return map.SelectMany(kvp => kvp.Value, (kvp, v) => new KeyValuePair<K, V>(kvp.Key, v)).GetEnumerator();
            }
            finally
            {
                if (_lock.IsReadLockHeld)
                {
                    _lock.ExitReadLock();
                }
            }
        }

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            throw new NotImplementedException();
        }
    }
}